flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/Demultiplexer.java

```diff
@@ -0,0 +1,402 @@
+					continue;
+				}
+			}
+
+			return result;
+			// loop is re-executed for filtered full records.
+		} while (true);
+	}
+
+	public void select(VirtualChannelSelector selector) {
+		selectedChannelIndex = selector.getChannelIndex();
+		selectedChannel = recordDeserializersForChannels.get(selectedChannelIndex);
+		if (selectedChannel == null) {
+			throw new IllegalStateException(
+				"Cannot select " + selector + "; known channels are " + getChannelSelectors().collect(Collectors.toList()));
+		}
+	}
+
+	@Override
+	public void setNextBuffer(Buffer buffer) throws IOException {
+		selectedChannel.setNextBuffer(buffer);
+	}
+
+	public void close() {
+		for (RecordDeserializer<DeserializationDelegate<StreamElement>> deserializer :
+			recordDeserializersForChannels.values()) {
+			// recycle buffers and clear the deserializer.
+			Buffer buffer = deserializer.getCurrentBuffer();
+			if (buffer != null && !buffer.isRecycled()) {
+				buffer.recycleBuffer();
+			}
+			deserializer.clear();
+		}
+	}
+
+	static ChannelDemultiplexer forChannel(
+			int subtaskIndex,
+			InputChannelInfo channelInfo,
+			DemultiplexParameters parameters,
+			int totalChannels) {
+		final InflightDataRescalingDescriptor mapping = parameters.channelMapping;
+		final RescaledChannelsMapping rescaledChannelsMapping =
+			mapping.getChannelMapping(channelInfo.getGateIdx());
+		final int[] oldChannels = rescaledChannelsMapping.getOldChannelIndexes(channelInfo.getInputChannelIdx());
+
+		final Map<Integer, Predicate<StreamRecord>> oldChannelsWithFilters =
+			Arrays.stream(oldChannels).boxed()
+				.collect(Collectors.toMap(
+					Function.identity(),
+					oldChannel -> getFilterForChannel(channelInfo, parameters, rescaledChannelsMapping, oldChannel)));
+
+		return new ChannelDemultiplexer(
+			subtaskIndex,
+			oldChannelsWithFilters,
+			parameters,
+			totalChannels);
+	}
+
+	private static Predicate<StreamRecord> getFilterForChannel(
+			InputChannelInfo channelInfo,
+			DemultiplexParameters parameters,
+			RescaledChannelsMapping rescaledChannelsMapping,
+			Integer oldChannel) {
+		return rescaledChannelsMapping.getNewChannelIndexes(oldChannel).length <= 1 ?
+			NO_FILTER :
+			createFilter(channelInfo, parameters);
+	}
+
+	private static Predicate<StreamRecord> createFilter(InputChannelInfo channelInfo, DemultiplexParameters parameters) {
+		final StreamPartitioner partitioner = parameters.gatePartitionerRetriever.apply(channelInfo.getGateIdx());
+		final int inputChannelIdx = channelInfo.getInputChannelIdx();
+		final SerializationDelegate<StreamRecord> delegate = parameters.delegate;
+		partitioner.setup(parameters.numberOfChannels);
+		if (partitioner instanceof ConfigurableStreamPartitioner) {
+			((ConfigurableStreamPartitioner) partitioner).configure(KeyGroupRangeAssignment.UPPER_BOUND_MAX_PARALLELISM);
+		}
+		return streamRecord -> {
+			delegate.setInstance(streamRecord);
```

Comment:
Since it's only used for the specific call inside the same thread it should be safe. However, it's cheap enough to create for each filter and it's probably much easier to reason about, so I changed it as you suggested.